addons/spreadsheet_account/models/account.py

```diff
@@ -79,50 +71,282 @@ def _build_spreadsheet_formula_domain(self, formula_params):
             domain = expression.AND(
                 [domain, [("move_id.state", "=", "posted")]]
             )
+        partner_ids = [int(partner_id) for partner_id in formula_params.get("partner_id", []) if partner_id]
+        if partner_ids:
+            domain = expression.AND(
+                [domain, [("partner_id", "in", partner_ids)]]
+            )
         return domain
 
+    def _pre_process_date_period_boundaries(self, args_list):
+        for args in args_list:
+            company_id = args['company_id'] or self.env.company.id
+            company = self.env['res.company'].browse(company_id)
+
+            start, _end = self._get_date_period_boundaries(
+                args['date_from'], company
+            )
+            _start, end = self._get_date_period_boundaries(
+                args['date_to'], company
+            )
+
+            args['date_from_boundary'] = start
+            args['date_to_boundary'] = end
+
+    def _pre_process_timeline(self, args_list):
+
+        # Get all boundaries
+        all_boundaries = set()
+        for args in args_list:
+            all_boundaries.add((args['date_from_boundary'], 'begin'))
+            all_boundaries.add((args['date_to_boundary'], 'end'))
+
+        all_boundaries = list(all_boundaries)
+        all_boundaries.sort()
+
+        # Compute non overlapping time period
+        timeline = []
+        for i in range(len(all_boundaries) - 1):
+            start = all_boundaries[i]
+            end = all_boundaries[i + 1]
+
+            timeline.append((
```

Comment:
I have no idea what you are trying to do, or what output you want to get. But I'm pretty sure that this is not it.
For instance, using
```python
        self.env['account.account'].spreadsheet_fetch_partner_balance([
            {
                'partner_ids': [self.move_23.partner_id.id],
                'date_from': {
                    'range_type': 'year',
                    'year': 2021,
                },
                'date_to': {
                    'range_type': 'year',
                    'year': 2021,
                },
                'codes': [],
                'company_id': None,
                'include_unposted': False,
            },
            {
                'partner_ids': [self.move_23.partner_id.id],
                'date_from': {
                    'range_type': 'year',
                    'year': 2022,
                },
                'date_to': {
                    'range_type': 'year',
                    'year': 2022,
                },
                'codes': [],
                'company_id': None,
                'include_unposted': False,
            },
            {
                'partner_ids': [self.move_23.partner_id.id],
                'date_from': {
                    'range_type': 'year',
                    'year': 2024,
                },
                'date_to': {
                    'range_type': 'year',
                    'year': 2024,
                },
                'codes': [],
                'company_id': None,
                'include_unposted': False,
            }
        ])
```

The timeline is
```python
|   timeline: [(datetime.date(2021, 1, 1), datetime.date(2021, 12, 31)),
               (datetime.date(2022, 1, 1), datetime.date(2021, 12, 31)),
               (datetime.date(2022, 1, 1), datetime.date(2022, 12, 31)),
               (datetime.date(2023, 1, 1), datetime.date(2023, 12, 31)),
               (datetime.date(2024, 1, 1), datetime.date(2024, 12, 31))]
```

Is that what you expected?